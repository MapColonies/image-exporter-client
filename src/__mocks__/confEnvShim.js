if (!window._env_) {
  window._env_ = (function (undefined) {
    return {
      LANGUAGE: 'en',
      MAP_SERVER: 'MAP_SERVER',
      PUBLISH_POINT: 'PUBLISH_POINT',
      CHANNEL: 1002,
      VERSION: 1,
      REQUEST: 'REQUEST',
      SERVICE_PROTOCOL: 'SERVICE_PROTOCOL',
      SERVICE_NAME: 'SERVICE_NAME',
      ACTIVE_LAYER: 'ACTIVE_LAYER',
      BBOX: {
        limit: 20000,
      },
      EXPORT: {
        defaultZoomLevel: 18,
        maxZoomLevel: 22
      },
      MAP: {
        center: [35.14, 31.39],
        zoom: 8,
      },
      ACTIVE_LAYER_PROPERTIES: {
        urlPattern : 'MAP_SERVER/service',
        exportUrlPattern: 'MAP_SERVER/service?REQUEST=REQUEST&SERVICE=SERVICE&LAYERS=LAYER',
        urlPatternParams: {
          service: 'SERVICE',
          layers: 'LAYER',
          tiled: 'TILED',
          matrixSet: 'MATRIXSET',
          style: 'STYLE',
          projection: 'PROJECTION'
        }
      },
    };
  })(void 0);
}
