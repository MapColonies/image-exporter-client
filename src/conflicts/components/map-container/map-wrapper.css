.ol-zoom {
  top: 2.5em;
  right: 0.5em;
  left: unset;
}

body[dir="rtl"] .ol-full-screen {
  right: unset;
  left: .5em;
}

body[dir="rtl"] .ol-zoom {
  right: unset;
  left: 0.5em;
}

.mdc-dialog__title + .mdc-dialog__content{
  padding-top: 20px;
}

.mdc-dialog__actions{
  padding-right: 24px;
}
