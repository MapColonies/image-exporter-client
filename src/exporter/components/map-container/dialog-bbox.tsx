import React, { useState } from 'react';
import * as  turf from '@turf/helpers';
import bbox from '@turf/bbox';
import bboxPolygon from '@turf/bbox-polygon';
import { Polygon } from 'geojson';
import { useFormik } from 'formik';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import {
  Dialog,
  DialogTitle,
  DialogContent,
  TextField,
  Button
} from '@map-colonies/react-core';
import { Box } from '@map-colonies/react-components';
import { FormattedMessage, useIntl, IntlShape } from 'react-intl';
import { BBoxCorner, Corner } from '../bbox/bbox-corner-indicator';
import { BBoxAreaLimit, isBBoxWithinLimit } from '../../../common/helpers/bbox-area';
import EXPORTER_CONFIG from '../../../common/config';

const useStyle = makeStyles((theme: Theme) =>
  createStyles({
    spacer: {
      marginRight: '16px'
    },
    errorContainer: {
      display: 'flex',
      alignItems: 'center',
      marginRight: 'auto',
      color: theme.palette.error.main,
    }

  })
);

interface BBoxCorners {
  bottomLeftLat: number;
  bottomLeftLon: number;
  topRightLat: number;
  topRightLon: number;
}

interface BBoxError {
  bboxArea: string;
}

const validate = (values: BBoxCorners, intl: IntlShape): BBoxError => {
  const errors: BBoxError = { bboxArea: '' };

  try {
    const line = turf.lineString([
      [values.bottomLeftLon, values.bottomLeftLat],
      [values.topRightLon, values.topRightLat],
    ]);
    const polygon = bboxPolygon(bbox(line));
    const isWithinLimit = isBBoxWithinLimit(polygon.geometry);
    if (isWithinLimit !== BBoxAreaLimit.OK) {
      const messageFormat = isWithinLimit === BBoxAreaLimit.TOO_BIG ? 'custom-bbox.form-error.area.large.text' : 'custom-bbox.form-error.area.small.text';
      errors.bboxArea = intl.formatMessage({ id: messageFormat });
    }
  } catch (err) {
    errors.bboxArea = intl.formatMessage({ id: 'custom-bbox.form-error.area.invalid.text' });
  }

  return errors;
};


interface DialogBBoxProps {
  isOpen: boolean;
  onSetOpen: (open: boolean) => void;
  onPolygonUpdate: (polygon: Polygon) => void;
}

export const DialogBBox: React.FC<DialogBBoxProps> = (
  props
) => {
  const { isOpen, onSetOpen, onPolygonUpdate } = props;
  const classes = useStyle();
  const intl = useIntl();
  const formik = useFormik({
    initialValues: {
      bottomLeftLat: 0,
      bottomLeftLon: 0,
      topRightLat: 0.00001,
      topRightLon: 0.00001,
    },
    onSubmit: values => {
      const err = validate(values, intl);

      if (!err.bboxArea) {
        const line = turf.lineString([
          [values.bottomLeftLon, values.bottomLeftLat],
          [values.topRightLon, values.topRightLat],
        ]);
        const polygon = bboxPolygon(bbox(line));

        onPolygonUpdate(polygon.geometry);
        handleClose(false);
        setFormErrors({
          bboxArea: ''
        });
      } else {
        setFormErrors(err);
      }

    },
  });

  const [formErrors, setFormErrors] = useState({ bboxArea: '' });

  const handleClose = (isOpened: boolean): void => {
    onSetOpen(isOpened);
  };

  // eslint-disable-next-line
  const checkValidLat = (e: React.ChangeEvent<any>) => {
    const maxMinLatValue = EXPORTER_CONFIG.MAP.MAX_MIN_LAT_VALUE;
    const isValid: boolean = isValidRegexPattern(e, maxMinLatValue);
    return isValid? formik.handleChange(e) : false;
  };

  // eslint-disable-next-line
  const checkValidLong = (e: React.ChangeEvent<any>) => {
    const maxMinLongValue = EXPORTER_CONFIG.MAP.MAX_MIN_LONG_VALUE;
    const isValid: boolean = isValidRegexPattern(e, maxMinLongValue);
    return isValid? formik.handleChange(e) : false;
  };

  const isValidRegexPattern = (e: React.ChangeEvent<any>, maxMinValue: number): boolean => {
    // eslint-disable-next-line
    const valueString: string = e.target.value as string;
    const splitChar = EXPORTER_CONFIG.EXPORT.SPLIT_CHARACTER;
    const maxFractionDigits = EXPORTER_CONFIG.EXPORT.MAX_FRACTION_DIGITS;
    const regexStr = `^([0-9])+\\${splitChar}?(\\b[0-9]{0,${maxFractionDigits}})*$`;
    const regex = new RegExp(regexStr);
    const value = Number(valueString);
    if (!valueString) {
      return false;
    }
    if (value < -maxMinValue || value > maxMinValue) {
      return false;
    }
    return regex.test(valueString);
  };

  return (
    <Dialog open={ isOpen } preventOutsideDismiss={ true }>
      <DialogTitle>
        <FormattedMessage id="custom-bbox.dialog.title" />
      </DialogTitle>
      <DialogContent>
        <form onSubmit={ formik.handleSubmit }>
          <Box style={ { display: 'flex', marginBottom: '16px' } }>
            <TextField
              label={ intl.formatMessage({ id: 'custom-bbox.dialog-field.top_right_lat.label' }) }
              id="topRightLat" 
              name="topRightLat"
              type="text"
              onChange={ checkValidLat }
              value={ formik.values.topRightLat }
              className={ classes.spacer }
            />
            <TextField
              label={ intl.formatMessage({ id: 'custom-bbox.dialog-field.top_right_lon.label' }) }
              id="topRightLon"
              name="topRightLon"
              type="text"
              onChange={ checkValidLong }
              value={ formik.values.topRightLon }
              className={ classes.spacer }
            />
            <BBoxCorner corner={ Corner.TOP_RIGHT } />
          </Box>
          <Box style={ { display: 'flex' } }>
            <TextField
              label={ intl.formatMessage({ id: 'custom-bbox.dialog-field.bottom_left_lat.label' }) }
              id="bottomLeftLat"
              name="bottomLeftLat"
              type="text"
              onChange={ checkValidLat }
              value={ formik.values.bottomLeftLat }
              className={ classes.spacer }
            />
            <TextField
              label={ intl.formatMessage({ id: 'custom-bbox.dialog-field.bottom_left_lon.label' }) }
              id="bottomLeftLon"
              name="bottomLeftLon"
              type="text"
              onChange={ checkValidLong }
              value={ formik.values.bottomLeftLon }
              className={ classes.spacer }
            />
            <BBoxCorner corner={ Corner.BOTTOM_LEFT } />
          </Box>
          <Box style={ { display: 'flex', justifyContent: 'flex-end', marginTop: '16px', gap: '16px' } }>
            {
              formErrors.bboxArea ?
                <div id="errorContainer" className={ classes.errorContainer }>
                  { `${ intl.formatMessage({ id: 'general.error.label' }) }: ${ formErrors.bboxArea }` }
                </div> :
                null
            }
            <Button type="button" onClick={ (): void => {
              handleClose(false);
            } }>
              <FormattedMessage id="general.cancel-btn.text" />
            </Button>
            <Button raised type="submit">
              <FormattedMessage id="general.ok-btn.text" />
            </Button>
          </Box>
        </form>
      </DialogContent>
    </Dialog>

  );
};
