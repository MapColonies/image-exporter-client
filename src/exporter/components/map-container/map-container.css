.map {
  height: 100%;
  width: 100%;
  position: fixed;
}

.filtersPosition {
  position: absolute;
  z-index: 1000;
  left: 12px;
  top: 12px;
}

body[dir="rtl"] .filtersPosition {
  position: absolute;
  z-index: 1000;
  left: unset;
  right: 12px;
  top: 12px;
}

.filtersContainer {
  display: flex;
  padding: 8px;
}

.filtersMargin {
  margin-left: 8px;
  display: flex;
  gap: 8px;
}

body[dir="rtl"] .filtersMargin {
  margin-right: 8px;
}
