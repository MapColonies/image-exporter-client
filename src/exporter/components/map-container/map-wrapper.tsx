import React from 'react';
import { Geometry } from 'geojson';
import rewind from '@turf/rewind';
import { Polygon } from 'geojson';
import { Map,
  VectorSource,
  GeoJSONFeature,
  VectorLayer,
  DrawInteraction,
  DrawType
 } from '@map-colonies/react-components';
import EXPORTER_CONFIG from '../../../common/config';
import './map-wrapper.css';

interface MapWrapperProps {
  drawType?: DrawType;
  selectionPolygon?: Polygon;
  onPolygonSelection: (polygon: Polygon) => void;
}

export const MapWrapper: React.FC<MapWrapperProps> = (props) => {
  const handlePolygonSelected = (geometry: Geometry): void => {
    const rewindedPolygon = rewind(geometry as Polygon);
    props.onPolygonSelection(rewindedPolygon);
  };

  return (
    <Map 
      allowFullScreen={true}
      showMousePosition={true}
      zoom={EXPORTER_CONFIG.MAP.ZOOM}
      center={EXPORTER_CONFIG.MAP.CENTER}
    >
      {props.selectionPolygon && (
        <VectorLayer>
          <VectorSource>
            <GeoJSONFeature geometry={props.selectionPolygon} />
          </VectorSource>
        </VectorLayer>
      )}
      {props.children}
      {props.drawType !== undefined && (
        <DrawInteraction
          drawType={props.drawType}
          onPolygonSelected={handlePolygonSelected}
        />
      )}
    </Map>
  );
};
