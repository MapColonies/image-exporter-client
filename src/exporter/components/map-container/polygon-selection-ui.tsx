import React from 'react';
import { Polygon } from 'geojson';
import { Menu, 
  MenuItem,
  Button,
  Tooltip } from '@map-colonies/react-core';
import '@map-colonies/react-core/dist/button/styles';
import '@map-colonies/react-core/dist/tooltip/styles';
import '@map-colonies/react-core/dist/menu/styles';

import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

import { Box, DrawType } from '@map-colonies/react-components';
import { useIntl, FormattedMessage } from 'react-intl';
import { DialogBBox } from './dialog-bbox';

const WIDTH_SPACING_FACTOR = 18;
const useStyle = makeStyles((theme: Theme) =>
  createStyles({
    drawingButton: {
      width: theme.spacing(WIDTH_SPACING_FACTOR),
    },
    fullWidth: {
      width: '100%',
      marginTop: '36px',
    },
  })
);

export interface PolygonSelectionUiProps {
  isSelectionEnabled: boolean;
  isDrawDisabled: boolean;
  onStartDraw: (type: DrawType) => void;
  onCancelDraw: () => void;
  onReset: () => void;
  onPolygonUpdate: (polygon: Polygon) => void;
}

export const PolygonSelectionUi: React.FC<PolygonSelectionUiProps> = (
  props
) => {
  const classes = useStyle();
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
  const intl = useIntl();
  const { isSelectionEnabled, onCancelDraw, onStartDraw, onReset, onPolygonUpdate } = props;

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>): void => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = (): void => {
    setAnchorEl(null);
  };

  const [open, setOpen] = React.useState(false);

  if (isSelectionEnabled) {
    return (
      <Tooltip 
        content={intl.formatMessage({id: 'polygon-selection.cancel-btn.tooltip'})} 
        align={'bottomLeft'}
      >
        <Button className={classes.drawingButton} raised onClick={onCancelDraw}>
          <FormattedMessage id="polygon-selection.cancel-btn.text"/>
        </Button>
      </Tooltip>
    );
  } else {
    return (
      <Box position="relative">
        <Tooltip
          content={intl.formatMessage({id: 'polygon-selection.draw-btn.tooltip'})}
          align={'bottomLeft'}
        >
          <Button
            className={classes.drawingButton}
            raised
            disabled={props.isDrawDisabled}
            onClick={handleClick}
          >
            <FormattedMessage id="polygon-selection.draw-btn.text"/>
          </Button>
        </Tooltip>
        <DialogBBox 
          isOpen={open}
          onSetOpen={setOpen}
          onPolygonUpdate={onPolygonUpdate}>
        </DialogBBox>
        <Menu
          className={classes.fullWidth}
          open={Boolean(anchorEl)}
          onClose={handleClose}
        >
          <MenuItem
            onClick={(): void => {
              onStartDraw(DrawType.BOX);
              handleClose();
            }}
          >
            <FormattedMessage id="polygon-selection.box-menu_option.text"/>
          </MenuItem>
          <MenuItem
            onClick={(): void => {
              setOpen(true);
            }}
          >
            <FormattedMessage id="polygon-selection.box_coorinate-menu_option.text"/>
          </MenuItem>
          <MenuItem
            onClick={(): void => {
              onReset();
              handleClose();
            }}
          >
            <FormattedMessage id="polygon-selection.clear-menu_option.text"/>
          </MenuItem>
        </Menu>
      </Box>
    );
  }
};
