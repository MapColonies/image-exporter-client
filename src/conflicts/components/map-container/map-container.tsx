import React, { useState } from 'react';
import { Polygon } from 'geojson';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import { DrawType } from  '@map-colonies/react-components'; 
import { PolygonSelectionUi } from './polygon-selection-ui';
import { MapWrapper } from './map-wrapper';

const PLACEMENT_SPACING_FACTOR = 1.5;
const WIDTH_SPACING_FACTOR = 80;
const useStyle = makeStyles((theme: Theme) =>
  createStyles({
    map: {
      height: '100%',
      width: '100%',
      position: 'fixed',
    },
    absolute: {
      position: 'absolute',
      zIndex: 1000,
      left: theme.spacing(PLACEMENT_SPACING_FACTOR),
      top: theme.spacing(PLACEMENT_SPACING_FACTOR),
    },
    filtersContainer: {
      display: 'flex',
      padding: theme.spacing(1),
    },
    filtersMargin: {
      marginLeft: theme.spacing(1),
    },
  })
);

export interface MapContainerProps {
  handlePolygonSelected: (polygon: Polygon) => void;
  handlePolygonReset: () => void;
  mapContent?: React.ReactNode;
  filters?: React.ReactNode[];
}

export const MapContainer: React.FC<MapContainerProps> = (
  props
) => {
  const [drawType, setDrawType] = useState<DrawType>();
  const [selectionPolygon, setSelectionPolygon] = useState<Polygon>();
  const classes = useStyle();

  const onPolygonSelection = (polygon: Polygon): void => {
    setSelectionPolygon(polygon);
    setDrawType(undefined);
    props.handlePolygonSelected(polygon);
  };

  const onReset = (): void => {
    setSelectionPolygon(undefined);
    props.handlePolygonReset();
  };

  return (
    <div className={classes.map}>
      <div className={`${classes.absolute}`}>
        <div className={classes.filtersContainer}>
          <PolygonSelectionUi
            onCancelDraw={(): void => setDrawType(undefined)}
            onReset={onReset}
            onStartDraw={setDrawType}
            isSelectionEnabled={drawType !== undefined}
            onPolygonUpdate={onPolygonSelection}
          />
          {props.filters?.map((filter, index) => (
            <div key={index} className={classes.filtersMargin}>
              {filter}
            </div>
          ))}
        </div>
      </div>
      <MapWrapper
        children={props.mapContent}
        onPolygonSelection={onPolygonSelection}
        drawType={drawType}
        selectionPolygon={selectionPolygon}
      />
    </div>
  );
};
