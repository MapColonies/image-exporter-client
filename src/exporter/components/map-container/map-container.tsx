import React, { useState } from 'react';
import { Polygon } from 'geojson';
import { DrawType } from  '@map-colonies/react-components';
import { BBoxAreaLimit, isBBoxWithinLimit } from '../../../common/helpers/bbox-area';
import { PolygonSelectionUi } from './polygon-selection-ui';
import { MapWrapper } from './map-wrapper';
import './map-container.css';

export interface MapContainerProps {
  selectionPolygon: Polygon;
  handlePolygonSelected: (polygon: Polygon) => void;
  handlePolygonReset: () => void;
  handleError: (isWithinLimit: BBoxAreaLimit) => void;
  isDrawDisabled: boolean;
  mapContent?: React.ReactNode;
  filters?: React.ReactNode[];
}

export const MapContainer: React.FC<MapContainerProps> = (
  props
) => {
  const [drawType, setDrawType] = useState<DrawType>();

  const onPolygonSelection = (polygon: Polygon): void => {
    const isWithinLimit = isBBoxWithinLimit(polygon);
    if(isWithinLimit !== BBoxAreaLimit.OK) {
      props.handleError(isWithinLimit);
    }

    setDrawType(undefined);
    props.handlePolygonSelected(polygon);
  };

  const onReset = (): void => {
    props.handlePolygonReset();
  };

  return (
    <div className="map">
      <div className="filtersPosition">
        <div className="filtersContainer">
          <PolygonSelectionUi
            onCancelDraw={(): void => setDrawType(undefined)}
            onReset={onReset}
            onStartDraw={setDrawType}
            isSelectionEnabled={drawType !== undefined}
            isDrawDisabled={props.isDrawDisabled}
            onPolygonUpdate={onPolygonSelection}
          />
          {props.filters?.map((filter, index) => (
            <div key={index} className="filtersMargin">
              {filter}
            </div>
          ))}
        </div>
      </div>
      <MapWrapper
        children={props.mapContent}
        onPolygonSelection={onPolygonSelection}
        drawType={drawType}
        selectionPolygon={props.selectionPolygon}
      />
    </div>
  );
};
