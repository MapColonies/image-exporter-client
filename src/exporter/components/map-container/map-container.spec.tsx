import React from 'react';
import { shallow } from 'enzyme';
// eslint-disable-next-line
import '../../../__mocks__/confEnvShim';
import { MapContainer } from './map-container';
import { MapWrapper } from './map-wrapper';


const polygonSelectedFn = jest.fn();
const polygonResetFn = jest.fn();
const DEFAULT_FILTERS_NUM =1;

describe('MapContainer component', () => {
  
  afterEach(() => {
    polygonSelectedFn.mockClear();
    polygonResetFn.mockClear();
  });

  it('renders correctly', () => {
    const wrapper = shallow(
      <MapContainer
        handlePolygonSelected={polygonSelectedFn}
        handlePolygonReset={polygonResetFn}
        filters={[]}
        mapContent={(<span>test</span>)}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('FILTERS has only ONE element when additional filters not passed', () => {
    const wrapper = shallow(
      <MapContainer
        handlePolygonSelected={polygonSelectedFn}
        handlePolygonReset={polygonResetFn}
        filters={undefined}
        mapContent={(<span>test</span>)}
      />
    );

    expect(wrapper.find('.filtersContainer').children()).toHaveLength(DEFAULT_FILTERS_NUM); 
  });

  it('transfered FILTERS added to container with default one', () => {
    const filtersArr = [<span>fliter1</span>, <span>fliter2</span>]
    const wrapper = shallow(
      <MapContainer
        handlePolygonSelected={polygonSelectedFn}
        handlePolygonReset={polygonResetFn}
        filters={filtersArr}
        mapContent={(<span>test</span>)}
      />
    );

    expect(wrapper.find('.filtersContainer').children()).toHaveLength(DEFAULT_FILTERS_NUM + filtersArr.length); 
  });

  it('MapContent prop elements transfered to MapWrapper component', () => {
    const wrapper = shallow(
      <MapContainer
        handlePolygonSelected={polygonSelectedFn}
        handlePolygonReset={polygonResetFn}
        filters={undefined}
        mapContent={(<span>test</span>)}
      />
    );

    expect(wrapper.find(MapWrapper).children()).toHaveLength(1); 
  });
});
