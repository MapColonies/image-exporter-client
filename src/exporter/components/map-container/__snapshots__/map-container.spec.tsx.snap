// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapContainer component renders correctly 1`] = `
<div
  className="map"
>
  <div
    className="filtersPosition"
  >
    <div
      className="filtersContainer"
    >
      <PolygonSelectionUi
        isSelectionEnabled={false}
        onCancelDraw={[Function]}
        onPolygonUpdate={[Function]}
        onReset={[Function]}
        onStartDraw={[Function]}
      />
    </div>
  </div>
  <MapWrapper
    onPolygonSelection={[Function]}
  >
    <span>
      test
    </span>
  </MapWrapper>
</div>
`;
